File: test/e2e/clients/python/framework/context.py
# Copyright 2025 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import uuid
import functools
import time
from typing import Optional, Dict
from test.e2e.clients.python.framework.predicates import (
    deployment_ready,
    pod_ready,
)

import kubernetes
import yaml


DEFAULT_KUBECONFIG_PATH = "bin/KUBECONFIG"


class TestContext:
    """Context for E2E tests, managing Kubernetes interactions"""

    def __init__(self, kubeconfig_path: Optional[str] = None):
        self.kubeconfig_path = kubeconfig_path or os.environ.get(
            "KUBECONFIG", DEFAULT_KUBECONFIG_PATH
        )
        self._api_client = None
        self.namespace = None

    def get_api_client(self):
        """Returns a Kubernetes API client"""
        if not self._api_client:
            self._api_client = kubernetes.config.new_client_from_config(
                self.kubeconfig_path
            )
        return self._api_client

    def get_core_v1_api(self):
        """Returns the CoreV1Api client"""
        return kubernetes.client.CoreV1Api(self.get_api_client())

    def get_apps_v1_api(self):
        """Returns the AppsV1Api client"""
        return kubernetes.client.AppsV1Api(self.get_api_client())

    def get_custom_objects_api(self):
        """Returns the CustomObjectsApi client"""
        return kubernetes.client.CustomObjectsApi(self.get_api_client())

    def create_temp_namespace(self, prefix="test-"):
        """Creates a temporary namespace for testing"""
        core_v1 = self.get_core_v1_api()
        namespace_name = f"{prefix}{uuid.uuid4().hex[:8]}"
        namespace_manifest = {
            "apiVersion": "v1",
            "kind": "Namespace",
            "metadata": {"name": namespace_name},
        }
        core_v1.create_namespace(body=namespace_manifest)
        self.namespace = namespace_name
        print(f"Created namespace: {self.namespace}")
        return self.namespace

    def delete_namespace(self, namespace: Optional[str] = None):
        """Deletes the specified namespace"""
        if namespace is None:
            namespace = self.namespace
        if namespace:
            core_v1 = self.get_core_v1_api()
            try:
                core_v1.delete_namespace(name=namespace)
                print(f"Deleted namespace: {namespace}")
                if self.namespace == namespace:
                    self.namespace = None
            except kubernetes.client.rest.ApiException as e:
                if e.status == 404:
                    print(f"Namespace {namespace} not found, skipping deletion.")
                else:
                    raise

    def apply_manifest_text(self, manifest_text: str, namespace: Optional[str] = None):
        """Applies the given manifest text to the cluster"""
        if namespace is None:
            namespace = self.namespace
        if not namespace:
            raise ValueError(
                "Namespace must be provided or created before applying manifests."
            )

        manifests = yaml.safe_load_all(manifest_text)
        for manifest in manifests:
            if not manifest:
                continue
            self._apply_single_manifest(manifest, namespace)

    def _apply_single_manifest(self, manifest: Dict, namespace: str):
        api_version = manifest.get("apiVersion")
        kind = manifest.get("kind")
        metadata = manifest.get("metadata", {})
        manifest_name = metadata.get("name")

        # Set namespace if not present
        if "namespace" not in metadata:
            manifest["metadata"]["namespace"] = namespace

        print(f"Applying {kind} '{manifest_name}' in namespace '{namespace}'")

        if api_version == "v1":
            core_v1 = self.get_core_v1_api()
            if kind == "Service":
                try:
                    core_v1.create_namespaced_service(
                        namespace=namespace, body=manifest
                    )
                except kubernetes.client.rest.ApiException as e:
                    if e.status == 409:  # Conflict
                        print(f"Service {manifest_name} already exists. Patching...")
                        core_v1.patch_namespaced_service(
                            name=manifest_name, namespace=namespace, body=manifest
                        )
                    else:
                        raise
        elif api_version == "apps/v1":
            apps_v1 = self.get_apps_v1_api()
            if kind == "Deployment":
                try:
                    apps_v1.create_namespaced_deployment(
                        namespace=namespace, body=manifest
                    )
                except kubernetes.client.rest.ApiException as e:
                    if e.status == 409:  # Conflict
                        print(f"Deployment {manifest_name} already exists. Patching...")
                        apps_v1.patch_namespaced_deployment(
                            name=manifest_name, namespace=namespace, body=manifest
                        )
                    else:
                        raise
        else:
            custom_objects_api = self.get_custom_objects_api()
            group, version = api_version.split("/")
            plural = self._get_plural_name(kind)  # Helper needed for this

            try:
                custom_objects_api.create_namespaced_custom_object(
                    group=group,
                    version=version,
                    namespace=namespace,
                    plural=plural,
                    body=manifest,
                )
            except kubernetes.client.rest.ApiException as e:
                if e.status == 409:  # Conflict
                    print(f"{kind} {manifest_name} already exists. Replacing...")
                    # Need to get the resource version for replace
                    try:
                        existing = custom_objects_api.get_namespaced_custom_object(
                            group=group,
                            version=version,
                            namespace=namespace,
                            plural=plural,
                            name=manifest_name,
                        )
                        manifest["metadata"]["resourceVersion"] = existing["metadata"][
                            "resourceVersion"
                        ]
                        custom_objects_api.replace_namespaced_custom_object(
                            group=group,
                            version=version,
                            namespace=namespace,
                            plural=plural,
                            name=manifest_name,
                            body=manifest,
                        )
                    except kubernetes.client.rest.ApiException as get_e:
                        print(f"Error getting existing object for replace: {get_e}")
                        raise
                else:
                    raise

    def _get_plural_name(self, kind: str) -> str:
        # Basic pluralization, may need refinement
        if kind == "SandboxWarmPool":
            return "sandboxwarmpools"
        if kind == "SandboxClaim":
            return "sandboxclaims"
        if kind == "SandboxTemplate":
            return "sandboxtemplates"
        if kind == "Sandbox":
            return "sandboxes"
        if kind.endswith("s"):
            return kind.lower() + "es"
        return kind.lower() + "s"

    def wait_for_object(
        self, watch_func, name: str, namespace: str, predicate_func, timeout=120
    ):
        """Waits for a Kubernetes object to satisfy a given predicate function"""
        w = kubernetes.watch.Watch()
        try:
            for event in w.stream(
                watch_func,
                namespace=namespace,
                field_selector=f"metadata.name={name}",
                timeout_seconds=timeout,
            ):
                obj = event["object"]
                if predicate_func(obj):
                    print(
                        f"Object {name} satisfied predicate on event type {event['type']}."
                    )
                    w.stop()
                    return True
            # Fallthrough means timeout
            raise TimeoutError(
                f"Object {name} did not satisfy predicate within {timeout} seconds."
            )
        except Exception as e:
            print(f"Error during watch: {e}")
            # Check if the error is due to a timeout in the stream
            if "timeout" in str(e).lower():
                raise TimeoutError(
                    f"Object {name} did not satisfy predicate within {timeout} seconds."
                )
            raise

    def wait_for_deployment_ready(
        self,
        name: str,
        namespace: Optional[str] = None,
        min_ready: int = 1,
        timeout=120,
    ):
        """Waits for a Deployment to have at least min_ready available replicas"""
        if namespace is None:
            namespace = self.namespace
        if not namespace:
            raise ValueError("Namespace must be provided.")

        apps_v1 = self.get_apps_v1_api()

        return self.wait_for_object(
            apps_v1.list_namespaced_deployment,
            name,
            namespace,
            deployment_ready(min_ready),
            timeout,
        )

    def wait_for_warmpool_ready(
        self,
        name: str,
        namespace: Optional[str] = None,
        min_ready: int = 1,
        timeout=120,
    ):
        """Waits for a SandboxWarmPool to have at least min_ready ready sandboxes"""
        if namespace is None:
            namespace = self.namespace
        if not namespace:
            raise ValueError("Namespace must be provided.")

        custom_objects_api = self.get_custom_objects_api()
        core_v1 = self.get_core_v1_api()
        is_pod_ready = pod_ready()

        try:
            warmpool = custom_objects_api.get_namespaced_custom_object(
                group="extensions.agents.x-k8s.io",
                version="v1alpha1",
                namespace=namespace,
                plural="sandboxwarmpools",
                name=name,
            )
            warmpool_uid = warmpool["metadata"]["uid"]
        except kubernetes.client.rest.ApiException as e:
            print(f"Error fetching SandboxWarmPool {name}: {e}")
            raise

        w = kubernetes.watch.Watch()
        ready_pods = set()
        try:
            for event in w.stream(
                core_v1.list_namespaced_pod,
                namespace=namespace,
                timeout_seconds=timeout,
            ):
                pod = event["object"]
                pod_name = pod.metadata.name

                owner_references = pod.metadata.owner_references or []
                is_owned = False
                for owner_ref in owner_references:
                    if (
                        owner_ref.kind == "SandboxWarmPool"
                        and owner_ref.uid == warmpool_uid
                    ):
                        is_owned = True
                        break

                if not is_owned:
                    continue

                event_type = event["type"]
                if event_type == "DELETED":
                    if pod_name in ready_pods:
                        ready_pods.remove(pod_name)
                    continue

                if is_pod_ready(pod):
                    ready_pods.add(pod_name)
                elif pod_name in ready_pods:
                    ready_pods.remove(pod_name)

                print(
                    f"WarmPool {name}: {len(ready_pods)}/{min_ready} pods ready. Current ready set: {ready_pods}"
                )
                if len(ready_pods) >= min_ready:
                    print(
                        f"SandboxWarmPool {name} is ready with {len(ready_pods)} pods."
                    )
                    w.stop()
                    return True

            # Fallthrough means timeout
            raise TimeoutError(
                f"SandboxWarmPool {name} did not become ready within {timeout} seconds."
            )
        except Exception as e:
            print(f"Error during watch: {e}")
            if "timeout" in str(e).lower():
                raise TimeoutError(
                    f"SandboxWarmPool {name} did not become ready within {timeout} seconds."
                )
            raise


if __name__ == "__main__":
    # Example Usage
    tc = None
    try:
        tc = TestContext()
        ns = tc.create_temp_namespace()

        print("TestContext example finished.")
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        if tc and tc.namespace:
            print(f"Cleaning up namespace: {tc.namespace}")
            # tc.delete_namespace()

File: test/e2e/clients/typescript/framework/context.ts
/*
Copyright 2025 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as crypto from "node:crypto";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { execFileSync } from "node:child_process";
import * as k8s from "@kubernetes/client-node";
import fetch, { type RequestInit } from "node-fetch";
import {
  deploymentReady,
  gatewayAddressReady,
  warmPoolReady,
} from "./predicates.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
// Project root is 5 levels up from test/e2e/clients/typescript/framework/
const PROJECT_ROOT = path.resolve(__dirname, "../../../../..");
const DEFAULT_KUBECONFIG_PATH = path.join(PROJECT_ROOT, "bin/KUBECONFIG");
const DEFAULT_TIMEOUT_SECONDS = 120;

/**
 * Context for E2E tests, managing Kubernetes interactions.
 */
export class TestContext {
  private kubeconfigPath: string;
  private _kubeConfig: k8s.KubeConfig | null = null;
  private _coreV1Api: k8s.CoreV1Api | null = null;
  private _appsV1Api: k8s.AppsV1Api | null = null;
  namespace: string | null = null;

  constructor(kubeconfigPath?: string) {
    this.kubeconfigPath = kubeconfigPath ?? process.env["KUBECONFIG"] ??
      DEFAULT_KUBECONFIG_PATH;
  }

  get kubeConfig(): k8s.KubeConfig {
    if (!this._kubeConfig) {
      this._kubeConfig = new k8s.KubeConfig();
      this._kubeConfig.loadFromFile(this.kubeconfigPath);
    }
    return this._kubeConfig;
  }

  get coreV1Api(): k8s.CoreV1Api {
    if (!this._coreV1Api) {
      this._coreV1Api = this.kubeConfig.makeApiClient(k8s.CoreV1Api);
    }
    return this._coreV1Api;
  }

  get appsV1Api(): k8s.AppsV1Api {
    if (!this._appsV1Api) {
      this._appsV1Api = this.kubeConfig.makeApiClient(k8s.AppsV1Api);
    }
    return this._appsV1Api;
  }

  /**
   * Creates a temporary namespace for testing.
   */
  async createTempNamespace(prefix: string = "test-"): Promise<string> {
    const namespaceName = `${prefix}${crypto.randomBytes(4).toString("hex")}`;
    await this.coreV1Api.createNamespace({
      body: {
        apiVersion: "v1",
        kind: "Namespace",
        metadata: { name: namespaceName },
      },
    });
    this.namespace = namespaceName;
    console.log(`Created namespace: ${this.namespace}`);
    return this.namespace;
  }

  /**
   * Deletes the specified namespace.
   */
  async deleteNamespace(namespace?: string): Promise<void> {
    const ns = namespace ?? this.namespace;
    if (!ns) return;

    try {
      await this.coreV1Api.deleteNamespace({ name: ns });
      console.log(`Deleted namespace: ${ns}`);
      if (this.namespace === ns) {
        this.namespace = null;
      }
    } catch (e: unknown) {
      const status = (e as { statusCode?: number }).statusCode;
      if (status === 404) {
        console.log(`Namespace ${ns} not found, skipping deletion.`);
      } else {
        throw e;
      }
    }
  }

  /**
   * Applies the given manifest text to the cluster using kubectl.
   */
  applyManifestText(manifestText: string, namespace?: string): void {
    const ns = namespace ?? this.namespace;
    if (!ns) {
      throw new Error(
        "Namespace must be provided or created before applying manifests.",
      );
    }

    try {
      const result = execFileSync(
        "kubectl",
        ["apply", "-f", "-", "-n", ns],
        {
          input: manifestText,
          encoding: "utf-8",
          stdio: ["pipe", "pipe", "pipe"],
          timeout: 30_000,
        },
      );
      if (result) {
        console.log(result);
      }
    } catch (e: unknown) {
      const err = e as { stderr?: string; status?: number };
      const errorMsg = err.stderr?.trim() ?? "unknown error";
      console.error(`kubectl apply failed: ${errorMsg}`);
      throw new Error(`Failed to apply manifest: ${errorMsg}`);
    }
  }

  /**
   * Waits for a Kubernetes object to satisfy a given predicate function using the Watch API.
   */
  async waitForObject<T>(
    watchPath: string,
    name: string,
    namespace: string,
    predicateFn: (obj: T) => boolean,
    timeout: number = DEFAULT_TIMEOUT_SECONDS,
  ): Promise<boolean> {
    const watcher = new k8s.Watch(this.kubeConfig);
    const startMs = Date.now();
    const timeoutMs = timeout * 1000;

    // Initial GET: check if already satisfied and capture resourceVersion to
    // avoid missing events that occur between the GET and watch establishment.
    let watchResourceVersion: string | undefined;
    let initialObj: (T & { metadata?: { resourceVersion?: string } }) | undefined;
    try {
      const cluster = this.kubeConfig.getCurrentCluster();
      if (cluster?.server) {
        const serverBase = cluster.server.replace(/\/$/, "");
        const fetchOpts = (await this.kubeConfig.applyToFetchOptions(
          {},
        )) as RequestInit;
        const timeoutSignal = AbortSignal.timeout(10_000);
        fetchOpts.signal = fetchOpts.signal
          ? AbortSignal.any([fetchOpts.signal as AbortSignal, timeoutSignal])
          : timeoutSignal;
        const resp = await fetch(
          `${serverBase}${watchPath}/${name}`,
          fetchOpts,
        );
        if (resp.ok) {
          initialObj = (await resp.json()) as T & {
            metadata?: { resourceVersion?: string };
          };
          watchResourceVersion = initialObj.metadata?.resourceVersion;
        } else {
          // Drain the body to release the TCP socket back to the keep-alive pool.
          resp.body?.resume();
        }
      }
    } catch {
      // Transient network/HTTP error — fall through to watch.
    }

    // Evaluate the predicate outside the network catch block so that
    // errors thrown by predicateFn propagate to the caller rather than
    // being silently swallowed as if they were transient GET failures.
    if (initialObj != null && predicateFn(initialObj)) {
      console.log(
        `Object ${name} already satisfied predicate (initial GET).`,
      );
      return true;
    }

    return new Promise<boolean>((resolve, reject) => {
      // Track whether the promise has been settled (resolved or rejected)
      // to prevent duplicate resolve/reject calls.
      let settled = false;
      let timedOut = false;

      let abortController: AbortController | undefined;
      let timer: ReturnType<typeof setTimeout>;

      const abortWatch = () => {
        if (abortController) {
          try {
            abortController.abort();
          } catch {
            // ignore
          }
        }
      };

      const cleanup = () => {
        clearTimeout(timer);
        abortWatch();
      };

      const elapsed = Date.now() - startMs;
      const remainingMs = Math.max(0, timeoutMs - elapsed);
      timer = setTimeout(() => {
        if (settled) return;
        settled = true;
        timedOut = true;
        cleanup();
        try {
          const resourceType = watchPath.split("/").pop() ?? "object";
          const desc = execFileSync(
            "kubectl",
            ["describe", resourceType, name, "-n", namespace],
            { encoding: "utf-8", stdio: ["ignore", "pipe", "pipe"], timeout: 5_000 },
          );
          const pods = execFileSync(
            "kubectl",
            ["get", "pods", "-n", namespace, "-o", "wide"],
            { encoding: "utf-8", stdio: ["ignore", "pipe", "pipe"], timeout: 5_000 },
          );
          console.error(
            `[waitForObject timeout ${name}] describe:\n${desc}`,
          );
          console.error(
            `[waitForObject timeout ${name}] pods:\n${pods}`,
          );
        } catch (dumpErr) {
          console.error(
            `[waitForObject timeout ${name}] dump failed:`,
            dumpErr,
          );
        }
        reject(
          new Error(
            `Object ${name} did not satisfy predicate within ${timeout} seconds.`,
          ),
        );
      }, remainingMs);

      watcher
        .watch(
          watchPath,
          {
            fieldSelector: `metadata.name=${name}`,
            // Pass the captured resourceVersion to ensure no events are missed
            // between the initial GET and watch establishment. Fall back to "0"
            // (replay from the API server's watch cache) when the GET failed or
            // returned a non-ok status — "0" ensures an ADDED event is delivered
            // for objects that already exist.
            resourceVersion: watchResourceVersion ?? "0",
          },
          (type: string, obj: T) => {
            console.log(
              `[watch ${name}] event=${type} status=${JSON.stringify((obj as { status?: unknown })?.status)}`,
            );
            if (settled) return;
            let matched: boolean;
            try {
              matched = predicateFn(obj);
            } catch (predicateErr) {
              // predicateFn threw — propagate to the caller. Without this guard
              // the k8s Watch library's own try/catch would silently discard the
              // exception, leaving the Promise to hang until the timeout fires.
              settled = true;
              cleanup();
              reject(predicateErr);
              return;
            }
            if (matched) {
              console.log(
                `Object ${name} satisfied predicate on event type ${type}.`,
              );
              settled = true;
              cleanup();
              resolve(true);
            }
          },
          (err?: unknown) => {
            cleanup();
            // Only reject if we haven't already settled.
            // When we abort the watch after success or timeout, the
            // doneCallback receives an abort error which we should ignore.
            if (err && !settled) {
              settled = true;
              reject(
                new Error(
                  `Watch error for ${name}: ${err instanceof Error ? err.message : String(err)
                  }`,
                ),
              );
            }
            if (!err && !settled) {
              settled = true;
              reject(
                new Error(
                  `Watch for ${name} ended before predicate was satisfied.`,
                ),
              );
            }
          },
        )
        .then((ac) => {
          abortController = ac;
          // Abort if already settled (timeout or predicate satisfied) before
          // the watch's AbortController was returned to us.
          if (timedOut || settled) {
            abortWatch();
          }
        })
        .catch((err) => {
          cleanup();
          if (!settled) {
            settled = true;
            reject(err);
          }
        });
    });
  }

  /**
   * Waits for a Deployment to have at least minReady available replicas.
   */
  async waitForDeploymentReady(
    name: string,
    namespace?: string,
    minReady: number = 1,
    timeout: number = DEFAULT_TIMEOUT_SECONDS,
  ): Promise<boolean> {
    const ns = namespace ?? this.namespace;
    if (!ns) throw new Error("Namespace must be provided.");

    return this.waitForObject<k8s.V1Deployment>(
      `/apis/apps/v1/namespaces/${ns}/deployments`,
      name,
      ns,
      deploymentReady(minReady),
      timeout,
    );
  }

  /**
   * Waits for a SandboxWarmPool to have all the required number of ready sandboxes.
   */
  async waitForWarmPoolReady(
    name: string,
    namespace?: string,
    timeout: number = DEFAULT_TIMEOUT_SECONDS,
  ): Promise<boolean> {
    const ns = namespace ?? this.namespace;
    if (!ns) throw new Error("Namespace must be provided.");

    return this.waitForObject<Record<string, any>>(
      `/apis/extensions.agents.x-k8s.io/v1beta1/namespaces/${ns}/sandboxwarmpools`,
      name,
      ns,
      warmPoolReady(),
      timeout,
    );
  }

  /**
   * Waits for a Gateway to have an address in its status.
   */
  async waitForGatewayAddress(
    name: string,
    namespace?: string,
    timeout: number = DEFAULT_TIMEOUT_SECONDS,
  ): Promise<boolean> {
    const ns = namespace ?? this.namespace;
    if (!ns) throw new Error("Namespace must be provided.");

    return this.waitForObject<Record<string, any>>(
      `/apis/gateway.networking.k8s.io/v1/namespaces/${ns}/gateways`,
      name,
      ns,
      gatewayAddressReady(),
      timeout,
    );
  }
}
